extensions/avro-extensions/src/main/java/io/druid/data/input/AvroStreamInputRowParser.java

```diff
@@ -0,0 +1,116 @@
+import io.druid.data.input.impl.TimestampSpec;
+import io.druid.data.input.schemarepo.SubjectAndIdConverter;
+import org.apache.avro.Schema;
+import org.apache.avro.generic.GenericDatumReader;
+import org.apache.avro.generic.GenericRecord;
+import org.apache.avro.io.DatumReader;
+import org.apache.avro.io.DecoderFactory;
+import org.apache.avro.util.ByteBufferInputStream;
+import org.joda.time.DateTime;
+import org.schemarepo.Repository;
+import org.schemarepo.api.TypedSchemaRepository;
+import org.schemarepo.api.converter.AvroSchemaConverter;
+
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.util.Collections;
+import java.util.List;
+
+public class AvroStreamInputRowParser<SUBJECT, ID> implements ByteBufferInputRowParser
+{
+  private final ParseSpec parseSpec;
+  private final List<String> dimensions;
+
+  private final TypedSchemaRepository<ID, Schema, SUBJECT> typedRepository;
+  private final SubjectAndIdConverter<SUBJECT, ID> subjectAndIdConverter;
+  private final Repository schemaRepository;
+
+  @JsonCreator
+  public AvroStreamInputRowParser(
+      @JsonProperty("parseSpec") ParseSpec parseSpec,
+      @JsonProperty("subjectAndIdConverter") SubjectAndIdConverter<SUBJECT, ID> subjectAndIdConverter,
+      @JsonProperty("schemaRepository") Repository schemaRepository
+  )
+  {
+    this.parseSpec = parseSpec;
+    this.dimensions = parseSpec.getDimensionsSpec().getDimensions();
+    this.subjectAndIdConverter = subjectAndIdConverter;
+    this.schemaRepository = schemaRepository;
+    typedRepository = new TypedSchemaRepository<ID, Schema, SUBJECT>(
+        schemaRepository,
+        subjectAndIdConverter.getIdConverter(),
+        new AvroSchemaConverter(false),
+        subjectAndIdConverter.getSubjectConverter()
+    );
+  }
+
+  @Override
+  public InputRow parse(ByteBuffer input)
+  {
+    ByteBufferInputStream inputStream = new ByteBufferInputStream(Collections.singletonList(input));
+    Pair<SUBJECT, ID> subjectAndId = subjectAndIdConverter.toSubjectAndId(input);
+    Schema schema = typedRepository.getSchema(subjectAndId.lhs, subjectAndId.rhs);
+    DatumReader<GenericRecord> reader = new GenericDatumReader<GenericRecord>(schema);
+    try {
+      GenericRecord record = reader.read(null, DecoderFactory.get().binaryDecoder(inputStream, null));
+      return doParse(record, parseSpec, dimensions);
+    }
+    catch (IOException e) {
+      throw new RuntimeException("Fail to decode avro message!", e);
+    }
+  }
+
+  protected static InputRow doParse(
+      GenericRecord record, ParseSpec parseSpec, List<String> dimensions
+  )
+  {
+    TimestampSpec timestampSpec = parseSpec.getTimestampSpec();
+    DateTime dateTime = timestampSpec.extractTimestamp(
+        Collections.singletonMap(
+            timestampSpec.getTimestampColumn(),
+            record.get(timestampSpec.getTimestampColumn())
+        )
+    );
+    return new MapBasedInputRow(dateTime, dimensions, new GenericRecordAsMap(record));
```

Comment:
@codingwhatever After checking out the AvroStorage wiki I got it, I just added a switch to let user can tell the hadoop parser to treat AvroStorage output array properly, please check if that is ok.

It's odd that in my tring to generate an AvroStorage output of avro array, the name of elements is "name_0" instead of "name", eg. input array: `{"someArray": [4, 9, 1]}`, AvroStorage output array: `{"someArray": [{"someArray_0": 4}, {"someArray_0": 9}, {"someArray_0": 1}]}`, didn't find why.

I did't follow the `genericRecordToMap` way, because I have several avro data whose schema is about 30 fields including nested maps, while my druid data source is only about 5 each for dimensions and metrics, so convert the whole data into map is a bit wasted(I didn't use an avro reader schema because there are 5 different not-well-designed schema in one druid data source, writting a common reader schema is kind of not easy). I'm wondering if there is someone else in the situation like me that want to save the slightly overhead of creating one `Map` for each avro record.